src/bpf/bpfctl.c

```diff
@@ -0,0 +1,238 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <getopt.h>
+
+#include "build.h"
+#include "format-table.h"
+#include "format-util.h"
+#include "main-func.h"
+#include "parse-argument.h"
+#include "pretty-print.h"
+#include "varlink.h"
+#include "verbs.h"
+
+static JsonFormatFlags arg_json_format_flags = JSON_FORMAT_OFF;
+
+typedef struct BpfProgramData BpfProgramData;
+typedef struct BpfMapData BpfMapData;
+
+struct BpfProgramData {
+        char *name;
+        uint64_t type;
+        uint64_t memlock;
+        char *tag;
+        char **map_names;
+        uint64_t run_time_ns;
+        uint64_t run_cnt;
+};
+
+struct BpfMapData {
+        char *name;
+        uint64_t type;
+        uint64_t memlock;
+};
+
+static void bpf_prog_data_done(BpfProgramData *p) {
+        assert(p);
+
+        p->name = mfree(p->name);
+        p->tag = mfree(p->tag);
+        strv_free(p->map_names);
+}
+
+static void bpf_map_data_done(BpfMapData *p) {
+        assert(p);
+
+        p->name = mfree(p->name);
+}
+
+static int query_bpf_progs_and_maps(void) {
+        _cleanup_(varlink_unrefp) Varlink *vl = NULL;
+        _cleanup_(table_unrefp) Table *map_table = NULL, *prog_table = NULL;
+        JsonVariant *i = NULL, *o = NULL, *m = NULL, *v = NULL;
+        int r;
+
+        map_table = table_new("map_name", "map_type", "memlock");
```

Comment:
hmm, these are supposed to be short but human-directed column names, i.e. should probably be "name" and "type" just.

Reply:
the "memlock" column, what's that? locked memory?

shouldn't this just be called "memory"?

Reply:
Hmm, the thing is, the line where this info comes from, in /proc/self/fdinfo/<fd>, itself labels the value as "memlock." So we might want to maintain consistency?

There is also the issue that this memlock value sometimes varies from the real memory usage (such as for non-preallocated bpf maps, which may increase or decrease real memory size dynamically), so labelling it as just 'memory' may be misleading.

If just 'memlock' is too vague though, I am open to renaming the column to 'locked memory' or something similar.

Reply:
so we generally want to expose a more "human" interface to people that procfs would do it. that said, if memlock is really the better term, then i am fine with it.